library/ssl_tls.c

```diff
@@ -1763,75 +1761,42 @@ static int ssl_decrypt_buf( mbedtls_ssl_context *ssl )
     ( defined(MBEDTLS_AES_C) || defined(MBEDTLS_CAMELLIA_C) )
     if( mode == MBEDTLS_MODE_CBC )
     {
-        /*
-         * Decrypt and check the padding
-         */
-        int ret;
-        unsigned char *dec_msg;
-        unsigned char *dec_msg_result;
-        size_t dec_msglen;
-        size_t minlen = 0;
-        size_t olen = 0;
-
-        /*
-         * Check immediate ciphertext sanity
-         */
-#if defined(MBEDTLS_SSL_PROTO_TLS1_1) || defined(MBEDTLS_SSL_PROTO_TLS1_2)
-        if( ssl->minor_ver >= MBEDTLS_SSL_MINOR_VERSION_2 )
-            minlen += ssl->transform_in->ivlen;
-#endif
-
-        if( ssl->in_msglen < minlen + ssl->transform_in->ivlen ||
-            ssl->in_msglen < minlen + ssl->transform_in->maclen + 1 )
```

Comment:
While I agree it's generally preferable to have length checks as close as possible to where we actually access that area of the buffer, in that specific case, I think it might be better to keep those checks here, before we have access to any secret data and are tempted to make decisions based on secret data.

(Also, from a performance / anti-DoS point of view, early reject before we did any significant work is also better.)

Reply:
I find this check hard to understand without further comment and prefer to do the checks when they are needed -- provided, of course, they do not constitute an information leak as the dangerous early return you pointed out above.